pennylane/estimator/resource_tracking.py

```diff
@@ -0,0 +1,383 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Core resource tracking logic."""
+from collections import defaultdict
+from collections.abc import Callable, Iterable
+from functools import singledispatch, wraps
+
+from pennylane.operation import Operation
+from pennylane.queuing import AnnotatedQueue, QueuingManager
+from pennylane.wires import Wires
+
+from .resource_config import ResourceConfig
+from .resource_operator import (
+    CompressedResourceOp,
+    GateCount,
+    ResourceOperator,
+)
+from .resources_base import DefaultGateSet, Resources
+from .wires_manager import Allocate, Deallocate, WireResourceManager
+
+# pylint: disable=protected-access,too-many-arguments
+
+
+def map_to_resource_op():  # TODO: Import this function instead when the mapping PR is merged
+    """Maps an instance of :class:`~.Operation` to its associated :class:`~.pennylane.labs.resource_estimation.ResourceOperator`."""
+
+
+def estimate(
+    obj: ResourceOperator | Callable | Resources | list,
+    gate_set: set | None = None,
+    zeroed: int = 0,
+    any_state: int = 0,
+    tight_budget: bool = False,
+    config: ResourceConfig | None = None,
+) -> Resources | Callable:
+    r"""Estimate the quantum resources required from a circuit or operation in terms of the gates
+    provided in the gateset.
+
+    Args:
+        obj (ResourceOperator | Callable | Resources | list): The quantum circuit or operation
+            to obtain resources from.
```

Comment:
Perhaps `workflow` is preferable to `obj`. Slightly more specific, although I'm not sure how much better we can do given the flexibility of input.
I believe we can remove the `list` option for now as the use case is unclear. It can always be added in the future if we We also want to emphasize `Callable` as the dominant option.
```suggestion
        workflow (Callable | ResourceOperator | Resources): The quantum circuit or operation
            for which to estimate resources.
```